nmoscommon/auth/auth_registrar.py

```diff
@@ -0,0 +1,200 @@
+
+API_NAMESPACE = 'x-nmos/auth/v1.0/'
+REGISTRATION_ENDPOINT = API_NAMESPACE + 'register_client'
+TOKEN_ENDPOINT = API_NAMESPACE + 'token'
+
+
+def get_credentials_from_file(file):
+    with open(CREDENTIALS_PATH, 'r') as f:
+        credentials = json.load(f)
+    client_id = credentials['client_id']
+    client_secret = credentials['client_secret']
+    return client_id, client_secret
+
+
+class AuthRegistrar(object):
+    def __init__(self, client_name, redirect_uri, client_uri=None,
+                 allowed_scope=None, allowed_grant="authorization_code",
+                 allowed_response="code", auth_method="client_secret_basic",
+                 certificate=None):
+        self.client_name = client_name
+        self.redirect_uri = redirect_uri
+        self.client_uri = client_uri
+        self.allowed_scope = allowed_scope
+        self.allowed_grant = allowed_grant
+        self.allowed_response = allowed_response
+        self.auth_method = auth_method
+        self.certificate = certificate  # Could be deployed centrally for an alternative trust mechanism
+
+        self.bridge = IppmDNSBridge()
+        self.initialised = self.initialise()
+        self.client_id = None
+        self.client_secret = None
+
+    def write_credentials_to_file(self, data, file_path):
+        self.client_id = data.get('client_id')
+        self.client_secret = data.get('client_secret')
+        credentials = {"client_id": self.client_id, "client_secret": self.client_secret}
+        with open(file_path, 'w') as f:
+            json.dump(credentials, f)
+        return True
+
+    def send_registration_request(self):
+        try:
+            href = self.bridge.getHref(MDNS_SERVICE_TYPE)
+            registration_href = urljoin(href, REGISTRATION_ENDPOINT)
+            print('Registration endpoint href is: {}'.format(registration_href))
+
+            data = {
+                    "client_name": self.client_name,
+                    "client_uri": self.client_uri,
+                    "scope": self.allowed_scope,
+                    "redirect_uri": self.redirect_uri,
+                    "grant_type": self.allowed_grant,
+                    "response_type": self.allowed_response,
+                    "token_endpoint_auth_method": self.auth_method
+                    }
+
+            # Decide how Basic Auth details are retrieved - user input? Retrieved from file?
+            reg_resp = requests.post(registration_href,
+                                     data=data,
+                                     auth=('dannym', 'password'),
+                                     timeout=0.5,
+                                     proxies={'http': ''}
+                                     )
+            reg_resp.raise_for_status()  # Raise error if status !=201
+            return reg_resp.json()
+        except HTTPError as e:
+            print("Unable to Register Client with Auth Server. {}".format(e))
+
+    def initialise(self):
```

Comment:
It feels like maybe this method should go under the constructor, as it's called from there? I dunno, was kinda where I was expecting to find it.